content/posts/2020-05-25-scaling-selfhosted-action-runners/index.md

```diff
@@ -0,0 +1,281 @@
+
+THe final step is to install the GitHub app we have created. Go to your [app settings](https://github.com/settings/apps/self-hosted-aws-spot-runners) and select install, now you can choose how to install the app. To only receive relevant events we choose to be repo specific and install the app to our test repo.
+
+![install-app](install-app.png 'Install app.')
+
+That's it, time to test. First create a simple action workflow in an empty directory.
+
+```bash
+mkdir -p .github/workflows
+cat << EOF > .github/workflows/verify.yml
+name: verify
+on: [push]
+
+jobs:
+  cowsay:
+    runs-on: self-hosted
+    container: npalm/cowsay
+
+    steps:
+      - uses: actions/checkout@v2
+      - run: cowsay -f ghostbusters "  Who ya gonna call?"
+EOF
+```
+
+Initialize your current directory as GitHub repo and push all the content to GitHub.
+
+```bash
+git init
+git remote add origin git@github.com:<YOUR_REPO>
+git add --all .github
+git commit -m "Test runners" .
+git push origin master
+```
+
+After 30 seconds you should see your new spot instance starting in AWS.
+
+![aws-running](aws-running.png)
+
+A bit later, the runner should be registered to your GitHub repo.
+
+![spot-runner](action-runners.png)
+
+And once registered it should only be a matter of seconds before the build starts.
+
+![workflow-run](run-workflow.png 'Workflow run')
+
+Every 5 minutes the scale down lambda is executed, and if you not running any job you should see the runner is terminated.
+
+![aws-shutting-down](aws-shutting-down.png)
+
+In case you build runners are not starting or registering. The best you can do is follow the trace. In the GitHub app on there is an advanced settings page where you can see the events events sent to the webhook. When the event is not accepted double check the endpoint and secret. Next you can check the logs for the webhook and scale up lambda in cloud watch. Finally you can inspect the EC2 user data logging. Access via SSM (MP: should this be ssh?) is by default enabled. Just select connect to the instance and inspect the log `/var/log/user_data.log`.
+
+## Final thoughts
+
+The abilities of GitHub self-hosted runners are arleady quit neat. But creating the setup as discussed requires some hacks and shortcuts. First of all it would be much easier once GitHub delivers the orchestration component, which will reduce the whole setup to only managing a bit of cloud resources. For example having runners on a Kubernetes cluster including scaling can make it a lot easier to use.
+
+By delivery an API for actions GitHub gives you the option to build your own setup, the fact that we always need to have a runner registered is quite an hack. Also, the way for scaling down will not win a beauty contest but is necessary because the API does not provide an option to verify wether a runner is still active. Furthermore, the cache of the distribution is bit of a hack to work around some crazy slow downloads that we had experienced.
```

Comment:
```suggestion
By delivery an API for actions GitHub gives you the option to build your own setup, the fact that we always need to have a runner registered is quite an hack. Also, the way for scaling down will not win a beauty contest but is necessary because the API does not provide an option to verify whether a runner is still active. Furthermore, the cache of the distribution is bit of a hack to work around some crazy slow downloads that we had experienced.
```